cmd/eval-dev-quality/cmd/evaluate.go

```diff
@@ -205,25 +214,45 @@ func (command *Evaluate) Initialize(args []string) (cleanup func()) {
 	{
 		commandRepositories := map[string]bool{}
 		commandRepositoriesLanguages := map[string]bool{}
-		for _, r := range command.Repositories {
-			languageIDOfRepository := strings.SplitN(r, string(os.PathSeparator), 2)[0]
-			commandRepositoriesLanguages[languageIDOfRepository] = true
-
-			if _, ok := languagesSelected[languageIDOfRepository]; ok {
-				commandRepositories[r] = true
-			} else {
-				command.logger.Printf("Excluded repository %s because its language %q is not enabled for this evaluation", r, languageIDOfRepository)
+		if len(command.Repositories) == 0 {
+			for _, language := range command.Languages {
+				commandRepositoriesLanguages[language] = true
+				commandRepositories[filepath.Join(language, evaluate.RepositoryPlainName)] = true
+
+				languagePath := filepath.Join(command.TestdataPath, language)
+				languageRepositories, err := os.ReadDir(languagePath)
+				if err != nil {
+					command.logger.Panicf("ERROR: language path %q cannot be accessed: %s", languagePath, err)
+				}
+
+				for _, repository := range languageRepositories {
+					if !repository.IsDir() {
+						continue
+					}
+					commandRepositories[filepath.Join(language, repository.Name())] = true
+				}
 			}
-		}
-		for languageID := range languagesSelected {
-			if len(command.Repositories) == 0 || commandRepositoriesLanguages[languageID] {
-				commandRepositories[filepath.Join(languageID, evaluate.RepositoryPlainName)] = true
-			} else {
-				command.Languages = slices.DeleteFunc(command.Languages, func(l string) bool {
-					return l == languageID
-				})
-				delete(languagesSelected, languageID)
-				command.logger.Printf("Excluded language %q because it is not part of the selected repositories", languageID)
+		} else {
+			for _, r := range command.Repositories {
+				languageIDOfRepository := strings.SplitN(r, string(os.PathSeparator), 2)[0]
+				commandRepositoriesLanguages[languageIDOfRepository] = true
+
+				if _, ok := languagesSelected[languageIDOfRepository]; ok {
+					commandRepositories[r] = true
+				} else {
+					command.logger.Printf("Excluded repository %s because its language %q is not enabled for this evaluation", r, languageIDOfRepository)
+				}
+			}
+			for languageID := range languagesSelected {
+				if commandRepositoriesLanguages[languageID] {
+					commandRepositories[filepath.Join(languageID, evaluate.RepositoryPlainName)] = true
+				} else {
+					command.Languages = slices.DeleteFunc(command.Languages, func(l string) bool {
+						return l == languageID
+					})
+					delete(languagesSelected, languageID)
+					command.logger.Printf("Excluded language %q because it is not part of the selected repositories", languageID)
+				}
```

Comment:
I find that a bit complicated. And we are having two different computation paths for basically the same situation: Assume the user does not specify a repository, that is, we do the `if` branch with the result that all repositories are evaluated. But if the user specifies all repositories, we do the `else` branch, again with the result that all repositories are evaluated. So we use two different paths for the same thing.

Ideally it should be something like this:

```go
if len(command.Repositories) == 0 {
  command.Repositories = AllRepositories
}

if len(command.Languages) == 0 {
  command.Languages == AllLanguages
}

if len(command.Tasks) == 0 {
  command.Tasks = AllTasks
}

ValidateOptions()
```

Reply:
Sure. For languages and tasks it is kinda straight forward (though the problem with languages is that we need to update them according to which repositories are selected).

The only problem is that `AllRepositories` means going through `testdata` and figuring out which ones there are (which is exactly what I am doing in the `if` case`). So yes we use two different paths for the same thing, but the first one is us reading the directory to figure out which repositories there are and the second one is just handling the users input, so two different scenarios.

Reply:
> though the problem with languages is that we need to update them according to which repositories are selected).

Should be the same with tasks at some point I think.

The difference between your appraoch an mine is that mine always performs option validation, and it's a separate step and easier to maintain IMO.
Your approach does not validate options if there are no repositories specified, because we do not have to yet. But once we have a better integration of tasks there needs to be a validation as well.

What I don't like about that code (not just the one you are changing) is that initialization and validation is mixed. I'd like to have validation in a separate function.

But okay, I guess it's fine for now. But please move the `if` branch into a funciton, and move `commandRepositories` and `commandRepositoriesLanguages` into the branches. They are not need afterwards AFAIS.